Schedule and Inventory access restrictions

If you create a volunteer account you can view data by going to these URLs
http://localhost:8080/schedule
http://localhost:8080/inventory

This data should only be accessible after an administrator has assigned the roles/inventory or roles/schedule to a volunteer.

I tried fixing this by adding the hasAuthorization method from foodController in the GET method as a parameter. It works but it is not passing the tests.

 1) Food Api User routes lists food categories:
     AssertionError: expected { message: 'User is not authorized' } to be an array
      at D:/pantry-for-good/server/tests/integration/food.spec.js:45:34
      at Test._assertFunction (node_modules\supertest\lib\test.js:281:11)
      at Test.assert (node_modules\supertest\lib\test.js:171:18)
      at Server.assert (node_modules\supertest\lib\test.js:131:12)
      at emitCloseNT (net.js:1558:8)
      at _combinedTickCallback (internal/process/next_tick.js:77:11)
      at process._tickCallback (internal/process/next_tick.js:104:9)

--- a/client/Router.js
+++ b/client/Router.js
@@ -2,7 +2,7 @@ import React from 'react'
 import {ConnectedRouter} from 'react-router-redux'
 import {Route} from 'react-router-dom'
 
-import {ADMIN_ROLE, clientRoles} from '../common/constants'
+import {ADMIN_ROLE, volunteerRoles} from '../common/constants'
 
 import Sidebar from './modules/core/components/sidebar/Sidebar'
 import Header from './modules/core/components/Header'
@@ -25,7 +25,9 @@ import requireRole from './components/router/requireRole'
 import SwitchWithNotFound from './components/router/SwitchWithNotFound'
 
 const IsAdmin = requireRole([ADMIN_ROLE])
-const IsVolunteer = requireRole([ADMIN_ROLE, clientRoles.VOLUNTEER])
+const canInventory = requireRole([ADMIN_ROLE, volunteerRoles.INVENTORY])
+const canPack = requireRole([ADMIN_ROLE, volunteerRoles.PACKING])
+const canSchedule = requireRole([ADMIN_ROLE, volunteerRoles.SCHEDULE])
 
 const Router = ({history}) =>
   <ConnectedRouter history={history}>
@@ -40,9 +42,9 @@ const Router = ({history}) =>
           <Route path="/customers" component={Customers} />
           <Route path="/donors" component={Donors} />
           <Route path="/drivers" component={Drivers} />
-          <Route path="/inventory" exact component={IsVolunteer(Inventory)} />
-          <Route path="/packing" exact component={IsVolunteer(Packing)} />
-          <Route path="/schedule" exact component={IsVolunteer(Schedule)} />
+          <Route path="/inventory" exact component={canInventory(Inventory)} />
+          <Route path="/packing" exact component={canPack(Packing)} />
+          <Route path="/schedule" exact component={canSchedule(Schedule)} />
           <Route path="/settings" component={IsAdmin(Settings)} />
           <Route path="/users" component={Users} />
           <Route path="/volunteers" component={Volunteers} />

--- a/server/controllers/food.js
+++ b/server/controllers/food.js
@@ -6,15 +6,20 @@ import {
   NotFoundError,
   ValidationError
 } from '../lib/errors'
-import {ADMIN_ROLE, volunteerRoles} from '../../common/constants'
+import {ADMIN_ROLE, clientRoles, volunteerRoles} from '../../common/constants'
 import Customer from '../models/customer'
 import Food from '../models/food'
+
+const {CUSTOMER} = clientRoles
+const {INVENTORY, SCHEDULE} = volunteerRoles
 
 export default {
   /**
    * Create a Food category
    */
   async create(req, res) {
+    authorizeByRole(req.user.roles, [INVENTORY])
+
     const food = new Food(req.body)
 
     const existingFoodCategory = await Food.find({'category': food.category, 'deleted': false}).lean()
@@ -31,6 +36,8 @@ export default {
    * Update a Food category
    */
   async update(req, res) {
+    authorizeByRole(req.user.roles, [INVENTORY])
+
     const food = extend(req.food, req.body)
 
     const savedFood = await food.save()
@@ -41,6 +48,8 @@ export default {
    * Delete a Food category
    */
   async delete(req, res) {
+    authorizeByRole(req.user.roles, [INVENTORY])
+
     const food = req.food
 
     // Prevent deleting if food category contains food items not marked as deleted
@@ -61,6 +70,8 @@ export default {
    * List of Food categories
    */
   async list(req, res) {
+    authorizeByRole(req.user.roles, [CUSTOMER, INVENTORY, SCHEDULE])
+
     const foods = await Food.find()
       .sort('category')
 
@@ -71,6 +82,8 @@ export default {
    * Create a food item
    */
   async createItem(req, res) {
+    authorizeByRole(req.user.roles, [INVENTORY])
+
     const item = req.body
     item.name = item.name.trim()
 
@@ -101,8 +114,13 @@ export default {
    * Update a food item
    */
   async updateItem(req, res) {
+    authorizeByRole(req.user.roles, [INVENTORY, SCHEDULE])
+
     const originalCategoryId = req.params.foodId
     const updatedItem = req.body
+
+    const involvesCategoryChange = !updatedItem.categoryId || updatedItem.categoryId === originalCategoryId
+    if (involvesCategoryChange) authorizeByRole(req.user.roles, [INVENTORY])
 
     const updatedCategory = await updateItemHelper(originalCategoryId, updatedItem)
     res.json(updatedCategory)
@@ -112,6 +130,8 @@ export default {
    * Delete a food item
    */
   async deleteItem(req, res) {
+    authorizeByRole(req.user.roles, [INVENTORY])
+
     const categoryId = req.food._id
     const foodItemId = req.itemId
     await Food.update({"_id": categoryId, "items._id": foodItemId}, { $set: {"items.$.deleted": true} })
@@ -137,19 +157,11 @@ export default {
   itemById(req, res, next, id) {
     req.itemId = id
     next()
-  },
+  }
+}
 
-  hasAuthorization(req, res, next) {
-    const authorizedRoles = [
-      ADMIN_ROLE,
-      volunteerRoles.INVENTORY,
-      volunteerRoles.PACKING,
-      volunteerRoles.SCHEDULE
-    ]
-
-    if (req.user && intersection(req.user.roles, authorizedRoles).length) {
-      return next()
-    }
+function authorizeByRole(userRoles, roles = []) {
+  if (!intersection(userRoles, [...roles, ADMIN_ROLE]).length) {
     throw new ForbiddenError
   }
 }

--- a/server/routes/food.js
+++ b/server/routes/food.js
@@ -20,17 +20,18 @@ export default () => {
 
   const foodRouter = Router({mergeParams: true})
 
+  foodRouter.use('/foods', requiresLogin)
   foodRouter.route('/foods')
-    .get(requiresLogin, foodController.list)
-    .post(foodController.hasAuthorization, websocketMiddleware(saveCatSchema), foodController.create)
+    .get(foodController.list)
+    .post(websocketMiddleware(saveCatSchema), foodController.create)
   foodRouter.route('/foods/:foodId')
-    .put(foodController.hasAuthorization, websocketMiddleware(saveCatSchema), foodController.update)
-    .delete(foodController.hasAuthorization, websocketMiddleware(deleteCatSchema), foodController.delete)
+    .put(websocketMiddleware(saveCatSchema), foodController.update)
+    .delete(websocketMiddleware(deleteCatSchema), foodController.delete)
   foodRouter.route('/foods/:foodId/items')
-    .post(foodController.hasAuthorization, websocketMiddleware(saveItemSchema), foodController.createItem)
+    .post(websocketMiddleware(saveItemSchema), foodController.createItem)
   foodRouter.route('/foods/:foodId/items/:itemId')
-    .put(foodController.hasAuthorization, websocketMiddleware(saveItemSchema), foodController.updateItem)
-    .delete(foodController.hasAuthorization, websocketMiddleware(deleteItemSchema), foodController.deleteItem)
+    .put(websocketMiddleware(saveItemSchema), foodController.updateItem)
+    .delete(websocketMiddleware(deleteItemSchema), foodController.deleteItem)
 
   // Finish by binding the food middleware
   foodRouter.param('foodId', foodController.foodById)

